### PR TITLE
Stateful map

### PR DESCRIPTION
**Pull request by MikeInnes**
See https://github.com/FluxML/Flux.jl/issues/1209.

It would be good if we could come up with a reasonable test case. It may be tricky to do that, and if so it may make sense to merge this to avoid the issues people are seeing.

### CODE REVIEW

**Comment by CarloLucibello** [comment]
should we merge this? the issue is sneaky and comes up quite often

**Comment by MikeInnes** [comment]
Yeah, we probably should just merge it and tag a patch.

**Comment by dhairyagandhi96** [comment]
bors r-

We should add tests for this case too

**Comment by CarloLucibello** [comment]
`reverse` has a mandatory `dims` keyword for multi-dimensional array. 
The easiest fix seems just to reverse vectors,  and leave higher-order arrays non-reversed

**Comment by AzamatB** [comment]
Bump, [Flux.jl#1209](https://github.com/FluxML/Flux.jl/issues/1209) is a really nasty and silent bug, can we please merge this fix?

File: src/lib/broadcast.jl [hunk 1]
```diff
@@ -154,6 +154,15 @@ end
   end
 end
 
+# # Use the `map` adjoint in this special case, which is the same but applies
+# # pullbacks in reverse order.
+# # This leaves regular `broadcast` technically incorrect when the broadcasted
+# # function is stateful.
+# # Look, I'm not proud of it, but this is extremely rare in practice.
```
Comment (DhairyaLGandhi) [inline src/lib/broadcast.jl:161]: Double comments

File: src/lib/array.jl [hunk 2]
```diff
@@ -179,8 +188,9 @@ for (mapfunc,∇mapfunc) in [(:map,:∇map),(:pmap,:∇pmap),(:vmap,:∇vmap)]
     else
       ys, backs = unzip(ys_and_backs)
       ys, function (Δ)
-        Δf_and_args_zipped = $mapfunc((f, δ) -> f(δ), backs, Δ)
-        Δf_and_args = unzip(Δf_and_args_zipped)
+        # Apply pullbacks in reverse order. Needed for correctness if `f` is stateful.
+        Δf_and_args_zipped = $mapfunc((f, δ) -> f(δ), _tryreverse($mapfunc, backs, Δ)...) 
+        Δf_and_args = unzip(_tryreverse($mapfunc, Δf_and_args_zipped))
```
Comment (DhairyaLGandhi) [inline src/lib/array.jl:193]: Any particular reason for the refactoring? Might be better to split out the adjoint for pmap etc to just do the sync and pmap in the pullback.

**Pull request merged**